### ISSUE TITLE
Unable to preserve mtime when extracting with tar onto xtreemfs

### ISSUE THREAD

**Issue by djohle** [xtreemfs/xtreemfs#427]
i've encountered and interesting behavior that is inconsistent with other filesystems, and I think it might be a bug in the way XtreemFS updates the modification times.

I have a file called nov11 with a timestamp of Nov 11:
# ls -l /tmp/nov11
-rw-r--r--. 1 root root 6 Nov 11 00:00 /tmp/nov11

I have a tar archive n11 with that file inside:
# tar tvf /tmp/n11
-rw-r--r-- root/root         6 2015-11-11 00:00 nov11

When I go to a local filesystem (type=xfs) an extract it in "preserve" mode I will get the follwoing:
# tar xpf /tmp/n11
# ls -l
-rw-r--r--. 1 root root 6 Nov 11 00:00 nov11

Now I'll do the same thing from within my XtreemFS mount point:
# tar xpf /tmp/n11 
# ls -l
-rw-r--r--. 1 root root 6 Nov 19 19:38 nov11

Ahh, the current date/time got set as the mtime for this file!  This is not right.

Digging a little deeper with `strace` I see that in both cases the system calls happened in the same order...

## On local filesystem:
[...]
openat(AT_FDCWD, "nov11", O_WRONLY|O_CREAT|O_EXCL|O_NOCTTY|O_NONBLOCK|O_CLOEXEC, 0600) = 4
write(4, "nov11\n", 6)                  = 6
dup2(4, 4)                              = 4
fstat(4, {st_mode=S_IFREG|0600, st_size=6, ...}) = 0
utimensat(4, NULL, {{1447982757, 871072428}, {1447221600, 0}}, 0) = 0
fchown(4, 0, 0)                         = 0
fchmod(4, 0644)                         = 0
close(4)                                = 0
[...]

## On XtreemFS filesystem:
[...]
openat(AT_FDCWD, "nov11", O_WRONLY|O_CREAT|O_EXCL|O_NOCTTY|O_NONBLOCK|O_CLOEXEC, 0600) = 4
write(4, "nov11\n", 6)                  = 6
dup2(4, 4)                              = 4
fstat(4, {st_mode=S_IFREG|0600, st_size=6, ...}) = 0
utimensat(4, NULL, {{1447982843, 0}, {1447221600, 0}}, 0) = 0
fchown(4, 0, 0)                         = 0
fchmod(4, 0644)                         = 0
close(4)                                = 0
[...]

And in both cases it did preserve the mtime (the number 1447221600 in utimensat() call is the 2015-11-11 timestamp).  So it must be the fchown() and/or fchmod() calls that are causing an update to the mtime on XtreemFS.

To test this theory, I prevented those two syscalls from happening:

# strace tar --no-same-permissions --no-same-owner -x -f /tmp/n11
[...]
openat(AT_FDCWD, "nov11", O_WRONLY|O_CREAT|O_EXCL|O_NOCTTY|O_NONBLOCK|O_CLOEXEC, 0644) = 4
write(4, "nov11\n", 6)                  = 6
dup2(4, 4)                              = 4
fstat(4, {st_mode=S_IFREG|0644, st_size=6, ...}) = 0
utimensat(4, NULL, {{1447983748, 0}, {1447221600, 0}}, 0) = 0
close(4)                                = 0
[...]

# ls -l
-rw-r--r--. 1 root root 6 Nov 19 19:42 nov11

Hey, look at that, it worked!  Confirmed, both of those calls are causing this (I did try them only one at a time as well, mtime got bumped up).

I don't know if it's a POSIX standard or whatnot, but from everything I've read and experienced, the rule is something like this...

ctime – File Change Time
ctime also changes when you change file's ownership or access permissions. It will also naturally highlight the last time file had its contents updated.

mtime – File Modify Time
Last modification time shows time of the  last change to file's contents. It does not change with owner or permission changes, and is therefore used for tracking the actual changes to data of the file itself.

It seems these definitions are not holding up under XtreemFS.

**Comment by djohle** [comment]
Curiously, merely changing the ownership or permissions direclty does not have this issue...

XtreemFS:
> host# chown root nov11
> host# chmod 600 nov11
> host# ls -l nov11
> -rw-------. 1 root root 0 Nov 11 00:00 nov11


The `strace` shows the key to that as the respective syscalls are:
> fchownat(AT_FDCWD, "nov11", 0, 4294967295, 0) = 0
and
> fchmodat(AT_FDCWD, "nov11", 0600)       = 0

**Comment by robert-schmidtke** [comment]
Thanks for the very detailed explanation. I could reproduce your problem with tar, but not with zip. We'll try and figure out what tar does differently from zip.

**Comment by robert-schmidtke** [comment]
Could you tell me which version of XtreemFS you are running?

**Comment by robert-schmidtke** [comment]
So I've narrowed down the issue a bit here with the following minimal example.

```c
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>

int main(int argc, char **argv) {
    struct timespec ts[2];
    ts[0].tv_sec = 1448033139;
    ts[0].tv_nsec = 0;
    ts[1].tv_sec = 1448009772;
    ts[1].tv_nsec = 0;

    int fd = open("f", O_CREAT | O_WRONLY | O_TRUNC, S_IRUSR | S_IWUSR);

    char buffer[8];
    strcpy(buffer, "Test.\n");
    
    write(fd, buffer, strlen(buffer));
    futimens(fd, ts);
    // utimensat(fd, NULL, ts, 0); // this should mean the same as above
    close(fd);
    // utimensat(AT_FDCWD, "f", ts, 0);
    return 0;
}
```

When you run this on your local file system, ```f``` has a timestamp Nov 20, 9:56. When you run it on XtreemFS, it has the current time. This is because the modification time gets reset upon close, which I think will need looking into: https://github.com/xtreemfs/xtreemfs/blob/master/cpp/src/libxtreemfs/volume_implementation.cpp#L487

When looking at the ```strace``` from ```tar``` we see that ```close``` is called after setting the timestamp. ```zip``` does not do that, which is why I guess I did not experience this issue with ```zip```.

```utimensat(fd, NULL, ts, 0)``` should do the same according to http://man7.org/linux/man-pages/man2/utimensat.2.html (NOTES), however it does not for me on my local file system (ext4). ```utimensat(AT_FDCWD, "f", ts, 0)``` after ```close``` works both on my local file system and XtreemFS.

**Comment by djohle** [comment]
I am using latest stable release: v1.5.1
All involved systems are running RHEL 7.2 (kernel 3.10.0-327.el7.x86_64).

For what it's worth, I did try ZIP as well and yes it does retain proper mtime.  Seems that is because ZIP calls utime() after the close() thus masking any update-mtime-on-close issue there may be.

Unzip to XtreemFS w/preserve:
> host# strace unzip -X /tmp/n11.zip 
> host# ls -l nov11
> -rw-r--r--. 1 root root 0 Nov 11 00:00 nov11

Relevant system calls from above:
> open("nov11", O_RDWR|O_CREAT|O_TRUNC, 0666) = 4
> fchmod(4, 0100644)                      = 0
> close(4)                                = 0
> utime("nov11", [2015/11/20-11:50:39, 2015/11/11-00:00:00]) = 0


I'm inclined to agree with you about the issue being related to the mtime getting set upon close.  Although the one counterexample is my 3rd tar strace output above (the one with the "no-same" options)) there is a utimensat() followed by close() yet it does retain the proper mtime somehow.

**Comment by robert-schmidtke** [comment]
You're right about your counterexample, I was thinking about the same thing and I could not figure out why. We'll come back with either an idea for a fix or an argument about why there's good reason for keeping it the way it is, especially because a lot more communication with the MRC would be involved in the solution (most likely), which is what we'd like to avoid for scalability reasons. Anyway I'd like to thank you again for the very specific directions and information you gave!
